File: nodejs/packages/periph/src/transport/sipo.js
'use strict';

const spi = require('spi-device');

/**
 * SiPo (serial-in/parallel-out shift register) transport for Node.js.
 *
 * Drives cascadable SIPO shift registers (TPIC6B595, SN74HC595, etc.) whose
 * SER IN/SRCK pins are electrically an SPI MOSI/SCK pair. Shifts data over
 * either a hardware spi-device or a bit-banged pair of opengpio Output objects.
 * RCK — and, if configured, SRCLR/G — are always plain opengpio Output objects,
 * independent of which SPI mode is used.
 *
 * Write-only: there is no read() or writeRead().
 *
 * Exactly one of {busNumber, deviceNumber} (hardware) or {serIn, srck}
 * (software) must be given in options.
 */
class SiPoTransport {
    /**
     * @param {object} rck - opengpio Output instance (register clock).
     * @param {object} [options]
     * @param {object} [options.srclr] - opengpio Output instance for SRCLR; omit to disable.
     * @param {object} [options.g] - opengpio Output instance for G (output enable); omit to disable.
     * @param {number} [options.busNumber] - Hardware mode: SPI bus number (opens spi-device).
     * @param {number} [options.deviceNumber] - Hardware mode: chip-select line on the bus.
     * @param {number} [options.maxSpeedHz=1000000] - Hardware mode: SPI clock in Hz.
     * @param {object} [options.serIn] - Software mode: opengpio Output instance for SER IN.
     * @param {object} [options.srck] - Software mode: opengpio Output instance for SRCK.
     */
    constructor(rck, options = {}) {
        const hardware = options.busNumber !== undefined;
        const software = options.serIn !== undefined;
        if (hardware === software) {
            throw new Error(
                'specify exactly one of {busNumber, deviceNumber} or {serIn, srck}');
        }

        this._rck = rck;
        this._srclr = options.srclr ?? null;
        this._g = options.g ?? null;
        this._serIn = options.serIn ?? null;
        this._srck = options.srck ?? null;

        if (hardware) {
            this._device = spi.openSync(options.busNumber, options.deviceNumber, {
                mode: spi.MODE0,
                maxSpeedHz: options.maxSpeedHz ?? 1_000_000,
            });
        } else {
            this._device = null;
        }

        this._rck.value = false;
        if (this._srclr !== null) this._srclr.value = true;
        if (this._g !== null) this._g.value = false;
    }

    /**
     * Shift data out MSB-first, then latch it into the output register.
     *
     * In hardware mode this transfers data over spi-device; in software mode
     * it bit-bangs SER IN/SRCK. Either way, RCK is then pulsed HIGH then LOW
     * to latch the shifted data into the storage register that drives the
     * outputs.
     *
     * @param {Buffer|Uint8Array} data - Bytes to shift out, one byte per cascaded device.
     */
    write(data) {
        const buf = Buffer.isBuffer(data) ? data : Buffer.from(data);
        if (this._device !== null) {
            this._device.transferSync([{ sendBuffer: buf, byteLength: buf.length }]);
        } else {
            for (const byte of buf) {
                for (let bit = 7; bit >= 0; bit--) {
                    this._serIn.value = !!((byte >> bit) & 1);
                    this._srck.value = true;
                    this._srck.value = false;
                }
            }
        }
        this._rck.value = true;
        this._rck.value = false;
    }

    /**
     * Pulse SRCLR LOW then HIGH to clear the shift register.
     *
     * The storage register (and therefore the outputs) is unaffected until
     * the next write().
     *
     * @throws {Error} If srclr was not configured.
     */
    clear() {
        if (this._srclr === null) throw new Error('SRCLR not configured');
        this._srclr.value = false;
        this._srclr.value = true;
    }

    /**
     * Drive G LOW (enabled) or HIGH (disabled).
     *
     * @param {boolean} enabled - true drives G LOW, letting the storage
     *   register drive the outputs. false drives G HIGH, forcing every
     *   output off without disturbing the storage register's contents.
     * @throws {Error} If g was not configured.
     */
    setOutputEnable(enabled) {
        if (this._g === null) throw new Error('G not configured');
        this._g.value = !enabled;
    }

    /**
     * Close the SPI device (if opened) and stop all configured GPIO pins.
     */
    close() {
        if (this._device !== null) this._device.closeSync();
        this._rck.stop();
        if (this._srclr !== null) this._srclr.stop();
        if (this._g !== null) this._g.stop();
        if (this._serIn !== null) this._serIn.stop();
        if (this._srck !== null) this._srck.stop();
    }
}

module.exports = { SiPoTransport };

File: rust/periph/src/transport/sipo.rs
//! SiPo (serial-in/parallel-out shift register) transport for chip drivers.
//!
//! Drives cascadable SIPO shift registers (TPIC6B595, SN74HC595, etc.) whose
//! SER IN/SRCK pins are electrically an SPI MOSI/SCK pair. The transport is
//! generic over any `embedded-hal` 1.0 [`SpiBus`] — a hardware bus or a
//! caller-supplied bit-banged `SpiBus` impl (e.g. from `embedded-hal-bus`)
//! are interchangeable, since the transport's type signature doesn't
//! distinguish them — plus [`OutputPin`] for RCK and `Option<impl OutputPin>`
//! for SRCLR/G.
//!
//! Write-only: there is no `read`/`write_read`.
//!
//! ## Usage
//!
//! ```rust,ignore
//! use periph::transport::sipo::SiPoTransport;
//!
//! let mut transport = SiPoTransport::new(spi, rck, Some(srclr), Some(g))?;
//! transport.write(&[0xA5])?;
//! transport.clear()?;
//! transport.set_output_enable(false)?;
//! ```
//!
//! ## Linux host (`linux-embedded-hal` crate)
//!
//! Use `linux_embedded_hal::SpidevBus` for hardware SPI, or any bit-banged
//! `SpiBus` impl (e.g. built on `linux_embedded_hal::CdevPin` via
//! `embedded-hal-bus`) for software SPI. `CdevPin` also provides RCK/SRCLR/G
//! in both modes.
//!
//! ```toml
//! linux-embedded-hal = "0.4"
//! embedded-hal = "1"
//! ```

use embedded_hal::digital::OutputPin;
use embedded_hal::spi::SpiBus;

/// Error type for [`SiPoTransport`] operations.
#[derive(Debug)]
pub enum SiPoError<SE, RE, CE, GE> {
    /// Error transferring data over the SPI bus.
    Spi(SE),
    /// Error driving the RCK pin.
    Rck(RE),
    /// Error driving the SRCLR pin.
    Srclr(CE),
    /// Error driving the G pin.
    G(GE),
    /// [`SiPoTransport::clear`] was called but SRCLR was not configured.
    SrclrNotConfigured,
    /// [`SiPoTransport::set_output_enable`] was called but G was not configured.
    GNotConfigured,
}

/// SiPo (serial-in/parallel-out shift register) transport.
///
/// Generic over any `embedded-hal` 1.0 [`SpiBus`] (SER IN/SRCK, hardware or
/// bit-banged) and [`OutputPin`] for RCK, with `Option<impl OutputPin>` for
/// SRCLR/G.
pub struct SiPoTransport<SPI, RCK, SRCLR, G> {
    spi: SPI,
    rck: RCK,
    srclr: Option<SRCLR>,
    g: Option<G>,
}

impl<SPI, RCK, SRCLR, G> SiPoTransport<SPI, RCK, SRCLR, G>
where
    SPI: SpiBus,
    RCK: OutputPin,
    SRCLR: OutputPin,
    G: OutputPin,
{
    /// Create a new transport: drive RCK LOW, SRCLR HIGH (if configured), and
    /// G LOW (if configured).
    ///
    /// # Arguments
    ///
    /// * `spi`   – Hardware or bit-banged SPI bus for SER IN/SRCK.
    /// * `rck`   – Pin connected to RCK (register clock).
    /// * `srclr` – Pin connected to SRCLR, or `None` to disable it.
    /// * `g`     – Pin connected to G (output enable), or `None` to disable it.
    pub fn new(
        spi: SPI,
        mut rck: RCK,
        mut srclr: Option<SRCLR>,
        mut g: Option<G>,
    ) -> Result<Self, SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        rck.set_low().map_err(SiPoError::Rck)?;
        if let Some(ref mut srclr) = srclr {
            srclr.set_high().map_err(SiPoError::Srclr)?;
        }
        if let Some(ref mut g) = g {
            g.set_low().map_err(SiPoError::G)?;
        }
        Ok(Self { spi, rck, srclr, g })
    }

    /// Shift `data` out MSB-first, then latch it into the output register.
    ///
    /// Transfers `data` over the SPI bus, then pulses RCK HIGH then LOW to
    /// latch the shifted data into the storage register that drives the
    /// outputs.
    ///
    /// # Arguments
    ///
    /// * `data` – Bytes to shift out, one byte per cascaded device.
    pub fn write(
        &mut self,
        data: &[u8],
    ) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        self.spi.write(data).map_err(SiPoError::Spi)?;
        self.rck.set_high().map_err(SiPoError::Rck)?;
        self.rck.set_low().map_err(SiPoError::Rck)?;
        Ok(())
    }

    /// Pulse SRCLR LOW then HIGH to clear the shift register.
    ///
    /// The storage register (and therefore the outputs) is unaffected until
    /// the next [`SiPoTransport::write`].
    ///
    /// # Errors
    ///
    /// Returns [`SiPoError::SrclrNotConfigured`] if SRCLR was not configured.
    pub fn clear(
        &mut self,
    ) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        let srclr = self.srclr.as_mut().ok_or(SiPoError::SrclrNotConfigured)?;
        srclr.set_low().map_err(SiPoError::Srclr)?;
        srclr.set_high().map_err(SiPoError::Srclr)?;
        Ok(())
    }

    /// Drive G LOW (`enabled = true`) or HIGH (`enabled = false`).
    ///
    /// `true` lets the storage register drive the outputs; `false` forces
    /// every output off without disturbing the storage register's contents.
    ///
    /// # Errors
    ///
    /// Returns [`SiPoError::GNotConfigured`] if G was not configured.
    pub fn set_output_enable(
        &mut self,
        enabled: bool,
    ) -> Result<(), SiPoError<SPI::Error, RCK::Error, SRCLR::Error, G::Error>> {
        let g = self.g.as_mut().ok_or(SiPoError::GNotConfigured)?;
        if enabled {
            g.set_low().map_err(SiPoError::G)?;
        } else {
            g.set_high().map_err(SiPoError::G)?;
        }
        Ok(())
    }

    /// Consume the transport and return the SPI bus and pins.
    pub fn release(self) -> (SPI, RCK, Option<SRCLR>, Option<G>) {
        (self.spi, self.rck, self.srclr, self.g)
    }
}
